specifyweb/frontend/js_src/lib/components/WorkBench/WbSpreadsheet.tsx

```diff
@@ -298,144 +280,190 @@ function WbSpreadsheetComponent({
+    if (dataset.rowresults && hot) {
       validation.getValidationResults();
+      workbench.cells.flushIndexedCellData = true;
     }
-  }, [dataset.rowresults, hotRef.current])
+  }, [dataset.rowresults, hot]);
+
+  // TODO: Move hottable props to a different file?
+  const autoWrapCol = React.useMemo(
+    () => userPreferences.get('workBench', 'editor', 'autoWrapCol'),
+    []
+  );
+  const autoWrapRow = React.useMemo(
+    () => userPreferences.get('workBench', 'editor', 'autoWrapRow'),
+    []
+  );
+  const columns = React.useMemo(
+    () =>
+      Array.from(
+        // Last column is invisible and contains disambiguation metadata
+        { length: dataset.columns.length + 1 },
+        (_, physicalCol) => ({
+          // Get data from nth column for nth column
+          data: physicalCol,
+        })
+      ),
+    []
+  );
+  const enterMoves = React.useMemo(
+    () =>
+      userPreferences.get('workBench', 'editor', 'enterMoveDirection') === 'col'
+        ? { col: 1, row: 0 }
+        : { col: 0, row: 1 },
+    []
+  );
+
+  const colHeaders = React.useCallback((physicalCol: number) => {
+    const tableIcon = mappings?.mappedHeaders?.[physicalCol];
+    const isMapped = tableIcon !== undefined;
+    const mappingCol = physicalColToMappingCol(physicalCol);
+    const tableName =
+      (typeof mappingCol === 'number'
+        ? mappings?.tableNames[mappingCol]
+        : undefined) ?? tableIcon?.split('/').slice(-1)?.[0]?.split('.')?.[0];
+    const tableLabel = isMapped
+      ? f.maybe(tableName, getTable)?.label ?? tableName ?? ''
+      : '';
+    // REFACTOR: use new table icons
+    return `<div class="flex gap-1 items-center pl-4">
+            ${
+              isMapped
+                ? `<img
+              class="w-table-icon h-table-icon"
+              alt="${tableLabel}"
+              src="${tableIcon}"
+            >`
+                : `<span
+              class="text-red-600"
+              aria-label="${wbPlanText.unmappedColumn()}"
+              title="${wbPlanText.unmappedColumn()}"
+            >${legacyNonJsxIcons.ban}</span>`
+            }
+            <span class="wb-header-name columnSorting">
+              ${dataset.columns[physicalCol]}
+            </span>
+          </div>`;
+  }, []);
+
+  const enterBeginsEditing = React.useMemo(
+    () => userPreferences.get('workBench', 'editor', 'enterBeginsEditing'),
+    []
+  );
+
+  const hiddenColumns = React.useMemo(() => {
+    return {
+      // Hide the disambiguation column
+      columns: [dataset.columns.length],
+      indicators: false,
+      // TODO: Typing possibly doesn't match for handsontable 12.1.0, fixed in 14
+      copyPasteEnabled: false,
+    };
+  }, []);
+
+  const hiddenRows = React.useMemo(() => {
+    return {
+      rows: [],
+      indicators: false,
+      // TODO: Typing possibly doesn't match for handsontable 12.1.0, fixed in 14
+      copyPasteEnabled: false,
+    };
+  }, []);
```

Comment:
(optional) if this is a constant object, you could assign it to a variable outside the component, and then just use it inside the component - that would remove the need for useMemo